src/lib.rs

```diff
@@ -0,0 +1,135 @@
+use crate::error::Error;
+use crate::html::{paper_wallets, to_data_url, WalletData};
+use bitcoin::{self, secp256k1, Address, AddressType, Network, PublicKey};
+use log::debug;
+use miniscript::bitcoin::secp256k1::SecretKey;
+use miniscript::bitcoin::PrivateKey;
+use miniscript::{self, Descriptor, DescriptorTrait, TranslatePk};
+use rand::{thread_rng, Rng};
+use std::collections::HashMap;
+
+mod error;
+mod html;
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug)]
+pub struct WifAndHexPub {
+    wif: String,
+    hex_pub: String,
+}
+
+/// Process descriptor and network and returns an html page encoded in a data-url which is pasteable
+/// in a browser
+pub fn process(descriptor: String, network: Network) -> Result<String> {
+    let descriptor_string: Descriptor<String> = descriptor.parse()?;
+    debug!("descriptor_string: {}", descriptor_string);
+
+    let (keys, address) = create_key_pairs_and_address(&descriptor_string, network)?;
+
+    let wallet_data = create_wallet_data(address, keys, &descriptor_string)?;
+
+    let html = paper_wallets(&wallet_data)?;
+
+    Ok(to_data_url(&html, "text/html"))
+}
+
+/// Creates a key pair for every alias in the given descriptor,
+/// Creates another descriptor replacing alias with the relative public key, so that we can compute the address.
+pub fn create_key_pairs_and_address(
+    descriptor_string: &Descriptor<String>,
+    network: Network,
+) -> Result<(HashMap<String, WifAndHexPub>, Address)> {
+    let mut rng = thread_rng();
+    let secp = secp256k1::Secp256k1::signing_only();
+    let mut keys = HashMap::new();
+    let alias_to_key = |alias: &String| -> Result<PublicKey> {
+        let entropy: [u8; 32] = rng.gen();
+        let sk = PrivateKey {
+            compressed: true,
+            network,
+            key: SecretKey::from_slice(&entropy)?,
+        };
+        let key = PublicKey::from_private_key(&secp, &sk);
+        keys.insert(
+            alias.clone(),
+            WifAndHexPub {
+                wif: sk.to_wif(),
+                hex_pub: key.to_string(),
+            },
+        );
+
+        Ok(key)
+    };
+    let descriptor_keys: Descriptor<PublicKey> =
+        descriptor_string.translate_pk(alias_to_key, |_| unreachable!())?;
```

Comment:
i would remove the `unreachable!()` here because i think relatively complex descriptors might trigger it. i think you could replace it with:

```rust
|alias| Ok(alias_to_key(alias)?.to_pubkeyhash())
```

that's probably made more complex by the fact that you are mutating `keys` from within the closure. in bdk i normally use two separate maps and i give a mut reference to one of the map to each closure call, and then i merge the content after the translation is done:

```rust
let mut map_a = HashMap::new();
let mut map_b = HashMap::new();

let closure = |alias: &String, map: &mut HashMap<_, _>| -> Result<PublicKey> {
  // ....
};

descriptor.translate_pk(|alias| closure(alias, &mut map_a), |alias| closure(alias, &mut map_b))?;

map_a.extend(map_b.drain());
```

alternatively i think you could also just wrap the map in a `RefCell` because i doubt `miniscript` internally uses multi-threading. so you could get away with it because it's guaranteed that only one of the two closures will borrow mutably the data at the same time.

Reply:
Removed the `unreachable!()` calls in e9dc9a4

even though it's not completely clear to me in which case they are called

Reply:
one example that triggers the panic is `wsh(and_v(vc:pk_h(K),pk(A)))`. not sure how likely it is for the compiler to emit that op-code (it's generally more efficient to just use the raw key), but still if somebody writes the descriptor by hand it could appear somewhere

Reply:
I see, thanks